### PR TITLE
Change to installer/helm to use peer authentication for global mTLS enable

### PR DESCRIPTION
**Pull request by diemtvu**
- For `global.mtls.enable=false`, we will not add any resource (neither using alpha nor beta API). For upgrade users, this means their existing policies in alpha API still working. For new user, the mesh is still in permissive mode as 1.5 default to permissive when no policy is set.

- For `global.mtls.enable=true`, we change nothing, so it's backward compatible to <1.4. In future, this mode will be removed completely.

### CODE REVIEW

**Comment by diemtvu** [comment]
@howardjohn  I've realized that I didn't commit the change in `installer` in the last PR. The change under `helm` is not necessary, though I think it's still better to make it consistent until it got removed.

Please take another look. Tkx

File: install/kubernetes/helm/istio/charts/security/templates/enable-mesh-mtls.yaml [hunk 1]
```diff
@@ -4,18 +4,21 @@
...
 
 # Authentication policy to enable mutual TLS for all services (that have sidecar) in the mesh.
-apiVersion: "authentication.istio.io/v1alpha1"
-kind: "MeshPolicy"
+apiVersion: "security.istio.io/v1beta1"
+kind: "PeerAuthentication"
 metadata:
   name: "default"
+  namespace: {{ .Release.Namespace }}
+spec:
```
Comment (howardjohn) [inline install/kubernetes/helm/istio/charts/security/templates/enable-mesh-mtls.yaml]: this looks wrong. 

Also, again, I don't think we should modify this folder. It is deprecated/frozen and no new changes are accepted

File: tests/e2e/tests/pilot/authn_policy_test.go [hunk 2]
```diff
@@ -21,23 +21,21 @@ import (
...
-		Kind: "MeshPolicy",
-		Name: "default",
-	}
-	// This policy will remove the permissive policy and enable mTLS mesh policy.
+	// Add global alpha mesh policy to enable mTLS. There should not be no beta nor alpha added
+	// in the test setup, so nothing need to be removed.
 	globalCfg := &deployableConfig{
 		Namespace:  "", // Use blank for cluster CRD.
 		YamlFiles:  []string{"testdata/authn/v1alpha1/global-mtls.yaml.tmpl"},
-		Removes:    []resource{globalPermissive},
```
Comment (howardjohn) [inline tests/e2e/tests/pilot/authn_policy_test.go]: These tests are only using the legacy helm stuff, which didn't change now? Do we still need them? Same applies to the dashboard

I am close to removing all of this by the way

Comment (diemtvu) [reply tests/e2e/tests/pilot/authn_policy_test.go]: Yes, we should remove them (I leave a TODO for that). Just don't want to make that change in this PR to make sure it doesn't break any other regression.

File: install/kubernetes/helm/istio/charts/security/templates/enable-mesh-mtls.yaml (hunk 1 above)
Comment (diemtvu) [reply install/kubernetes/helm/istio/charts/security/templates/enable-mesh-mtls.yaml]: Removed.

File: tests/e2e/tests/pilot/authn_policy_test.go (hunk 2 above)
Comment (howardjohn) [reply tests/e2e/tests/pilot/authn_policy_test.go]: But nothing in this PR impacts this at all. the tests under tests/e2e use install/, not manifests/, so I don't see why we need to change anything? Maybe I am missing something though

Comment (diemtvu) [reply tests/e2e/tests/pilot/authn_policy_test.go]: Hmm. May be you are right. I saw the error in the first presubmit run when I made the changes in both. I can try to revert to see what happens

**Comment by diemtvu** [comment]
Note: pending on https://github.com/istio/istio/pull/20881

both integ-security-k8s-tests_istio and integ-telemetry-k8s-tests_istio fails because auto-mTLS hasn't been implemented for beta API. As a result, principal does not exist for authZ-deny check, nor metadata for telemtry.

File: tests/e2e/tests/dashboard/dashboard_test.go [hunk 3]
```diff
@@ -248,6 +248,21 @@ func replaceGrafanaTemplates(orig string) string {
 	return queryCleanupReplacer.Replace(out)
 }
 
+func istioQueryFilterFn(queries []string) []string {
```
Comment (howardjohn) [inline tests/e2e/tests/dashboard/dashboard_test.go]: same comment as the other test, this shouldn't be impacted

**Review by howardjohn — APPROVED** [review]

**Comment by diemtvu** [comment]
It does. As we no longer add the mesh policy during installation, this
metrics query return nill.

We will have different PR to change the dashboard to reflect the new policy
later.

**Comment by howardjohn** [comment]
> It does. As we no longer add the mesh policy during installation, this

This test uses the legacy helm though.. it shouldn't be impacted

**Review by costinm — DISMISSED** [review]
This is very dangerous - please don't add the API to the installer. It is ok if user is explicitly setting it,
but should not be set in the installer. 

There are many install/upgrade paths - in the past we had jobs to make sure old resources were deleted ( in case users didn't use --prune), and this impacts our ability to ship multicontrol planes in 1.6 - since this will apply to all control planes.

**Comment by diemtvu** [comment]
> This is very dangerous - please don't add the API to the installer. It is ok if user is explicitly setting it,
> but should not be set in the installer.
> 
> There are many install/upgrade paths - in the past we had jobs to make sure old resources were deleted ( in case users didn't use --prune), and this impacts our ability to ship multicontrol planes in 1.6 - since this will apply to all control planes.

I actually want to get rid of that too (it was there since ever, for the old API + DR). We can remove this `global.mtl.enabled` completely, and let user drive their settings via API. It will touch a bunch of e2e tests though

**Comment by costinm** [comment]
Removing the old option is also unsafe - may break security for users.

Please leave the old one in place, in 1.6 we can remove it after we have a migration tool (as is required for any such change).

1.6 control planes will then ignore the old setting, and the new one will be under user control - as it should be for any API. 

There are subtle cases - user may start with permissive and create a (old) authn config by himself. 
All of them have critical risks on user security.

The feature should not be on by default as it's launched anyways - and please make sure there is 
a flag that allows quick disabling of the new code.

**Comment by diemtvu** [comment]
> Removing the old option is also unsafe - may break security for users.
> 
> Please leave the old one in place, in 1.6 we can remove it after we have a migration tool (as is required for any such change).
> 
> 1.6 control planes will then ignore the old setting, and the new one will be under user control - as it should be for any API.
> 
> There are subtle cases - user may start with permissive and create a (old) authn config by himself.
> All of them have critical risks on user security.
> 
> The feature should not be on by default as it's launched anyways - and please make sure there is
> a flag that allows quick disabling of the new code.

I mean remove the --set values.global.mtls.enabled=x option. We will *not* adding any resource to config mTLS at installation time.

The implementation already fallback to alpha policy if available, so that strategy works fine with upgrading scenario. Alpha policies still take effect in that case.

For new installation and going forward, it's clearer for users, though they may need to set the policy explicitly. On the other hand, we don't have to handle the problem of adding resources at installation.

The switch for this API is simply remove it.

**Comment by diemtvu** [comment]
> Removing the old option is also unsafe - may break security for users.
> 
> Please leave the old one in place, in 1.6 we can remove it after we have a migration tool (as is required for any such change).
> 
> 1.6 control planes will then ignore the old setting, and the new one will be under user control - as it should be for any API.
> 
> There are subtle cases - user may start with permissive and create a (old) authn config by himself.
> All of them have critical risks on user security.
> 
> The feature should not be on by default as it's launched anyways - and please make sure there is
> a flag that allows quick disabling of the new code.

FYI, I have test the upgrade command (in permissive mode) on 1.4 mesh. It does not remove the existing `MeshConfig` as expected. We do not change the behavior on strict mode.

I still think this change makes a cleaner installation for 1.5. The default upgrade/installation add no policy (and as confirmed above, it doesn't prune the existing), so doesn't make any change to the existing mesh. At the same time, it will help avoid confusion of what API is used in the newly create mesh.

Thus I'd like to have this submitted and cherry picked to 1.5. What do you think?

File: install/kubernetes/helm/istio/charts/security/templates/enable-mesh-permissive.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+{{- define "security-permissive.yaml.tpl" }}
```
Comment (costinm) [inline install/kubernetes/helm/istio/charts/security/templates/enable-mesh-permissive.yaml]: Where is this used ? Permissive should be the default, no need to add it.

Comment (diemtvu) [reply install/kubernetes/helm/istio/charts/security/templates/enable-mesh-permissive.yaml]: I forgot to rebased. This file was deleted in master. Should be gone now.

**Pull request merged**